wit-0.3.0-draft/stdio.wit

```diff
@@ -1,17 +1,59 @@
+@since(version = 0.3.0-rc-2025-08-15)
+interface stdio {
+  @since(version = 0.3.0-rc-2025-08-15)
+  enum error-code {
+    /// Input/output error
+    io,
+    /// Invalid or incomplete multibyte or wide character
+    illegal-byte-sequence,
+    /// Broken pipe
+    pipe,
+  }
+}
+
 @since(version = 0.3.0-rc-2025-08-15)
 interface stdin {
+  /// Return a stream for reading from stdin.
+  ///
+  /// This function returns a stream which provides data read from stdin,
+  /// and a future to signal read results.
+  ///
+  /// If the stream's readable end is dropped the future will resolve to success.
```

Comment:
I'm not sure if this is a good idea. If an implementation received an async SIGPIPE or something it might want to drop its end and resolve the future without waiting for the reader to poll?

Reply:
We're modeling eventual `stream<T, U, V>` here and in that use case I think we would only ever get a `U` "finalizer" element if `stream<T>` is exhausted.
Perhaps we should expand a bit in `wasi:filesystem` docs about this as well https://github.com/WebAssembly/wasi-filesystem/blob/d81d6256c271fe1c8937eb8353e2ddc25517c153/wit-0.3.0-draft/types.wit#L300-L315, but my understanding of this pattern so far was that the `future` is *only* guaranteed to resolve once the write end of the stream is dropped.

That is, implementations will get a `dropped` reading from the stream and only then will poll the future to determine if the read of the stream was successful or not.
Polling the future *before* the read stream has been exhausted is not guaranteed to resolve at all

Reply:
Here's what I'm thinking:

- Guest calls `read-stdin`
- Host buffer is empty, so it does ~`spawn_blocking(stdin.read())`, returns pending
- Guest awaits stream along with some other waitables
- In _parallel_:
  - Guest gets some other event, decides it doesn't care about stdin right now; drops the stream and future
  - Host `stdin.read` fails with EPIPE, which it writes to the future

In this case, the EPIPE is just lost, right? i.e. if the guest calls `read-stdin` again what will it get?

Reply:
It's not possible for the guest to drop the stream it's reading from without cancelling the read in progress. With that the host would then be responsible for properly propagating that cancellation signal. If during cancellation `EPIPE` happened the host could report "ok you cancelled but the thing finished with `dropped` anyway" and then the guest gets `EPIPE` delivered too. If the host cancels before `EPIPE` then that's the host's probably to buffer that for another time.

Basically due to the explicitly requirement of cancelling the active read I think the situation you're thinking about @lann is resolved

Reply:
Oh? The [implementation right now](https://github.com/bytecodealliance/wasmtime/blob/4e8ab840ca9c07f59b19aa034c0be267c1f410f9/crates/wasi/src/p3/filesystem/host.rs#L164-L166) ignores cancellation basically meaning that if it's cancelled you're told that the cancellation is in progress or you're blocked until that completes. Could you elaborate on how read-via-stream is different?

Reply:
I've better wrapped my ahead around this now and see what's going on. :+1: